src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory =
+                options.CacheDirectory ??
+                throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetCachedEnvelopeFiles()
+        {
+            if (!_cacheDirectory.Exists)
+            {
+                return Enumerable.Empty<FileInfo>();
+            }
+
+            return _cacheDirectory.EnumerateFiles("*.envelope");
+        }
+
+        private FileInfo? TryGetNextEnvelopeFile() => GetCachedEnvelopeFiles()
+            .OrderBy(f => f.CreationTimeUtc)
+            .FirstOrDefault();
+
+        private async ValueTask FlushCacheAsync(CancellationToken cancellationToken = default)
+        {
+            while (TryGetNextEnvelopeFile() is { } envelopeFile)
+            {
+                Options.DiagnosticLogger?.LogDebug("Sending cached envelope: {0}", envelopeFile.FullName);
+
+                using var envelope = await Envelope.DeserializeAsync(
+                    envelopeFile.OpenRead(),
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
```

Comment:
We default to deleting the file. If there was an exception while trying to send it, for example. Or if Sentry returned 400, 500.

The exception is when we can clearly identify we couldn't connect to Sentry (timeout, or offline).
In this case we keep the file around to try again later.

That said I believe we're missing a mechanism to communicate that here. And some try/finally(delete)

Reply:
That's true, I will update it